tegra/tegra.c

```diff
@@ -57,25 +57,158 @@ static struct drm_tegra_bo * lookup_bo(void *table, uint32_t key)
 	return bo;
 }
 
+static void drm_tegra_bo_setup_guards(struct drm_tegra_bo *bo)
+{
+	struct drm_tegra *drm = bo->drm;
+	struct drm_tegra_gem_mmap args;
+	uint64_t guard = 0x5351317315731757;
+	uint8_t *map;
+	size_t size;
+	unsigned i;
+	int err;
+
+	if (!drm->debug_bo_front_guard && !drm->debug_bo_back_guard)
+		return;
+
+	size = bo->size;
+
+	if (drm->debug_bo_back_guard)
+		size += 4096;
+
+	memset(&args, 0, sizeof(args));
+	args.handle = bo->handle;
+
+	err = drmCommandWriteRead(drm->fd, DRM_TEGRA_GEM_MMAP,
+				  &args, sizeof(args));
+	if (err < 0) {
+		VDBG_BO(bo, "failed get mapping offset err %d (%s)\n",
+			err, strerror(-err));
+		abort();
+	}
+
+	map = mmap(0, bo->offset + size, PROT_READ | PROT_WRITE, MAP_SHARED,
+		   drm->fd, args.offset);
+	if (map == MAP_FAILED) {
+		VDBG_BO(bo, "failed to map guard 0x%llX err %d (%s)\n",
+			args.offset, -errno, strerror(errno));
+		abort();
+	}
+
+	if (drm->debug_bo_front_guard)
+		bo->guard_front = (uint64_t *)map;
+
+	if (drm->debug_bo_back_guard)
+		bo->guard_back = (uint64_t *)(map + bo->offset + bo->size);
+
+	VDBG_BO(bo, "front %p back %p\n", bo->guard_front, bo->guard_back);
+
+	/* we only interested in guards mapping, hence unmap the actual BO */
+	if (bo->size >= 4096)
+		munmap(map + bo->offset, align(bo->size - 4095, 4096));
+
+	if (bo->handle & 1)
+		guard = ~guard;
+
+	if (drm->debug_bo_front_guard)
+		for (i = 0; i < 512; i++)
+			bo->guard_front[i] = guard;
+
+	if (drm->debug_bo_back_guard)
+		for (i = 0; i < 512; i++)
+			bo->guard_back[i] = guard;
+}
+
+static void drm_tegra_bo_check_guards(struct drm_tegra_bo *bo)
+{
+	struct drm_tegra *drm = bo->drm;
+	uint64_t guard = 0x5351317315731757;
+	unsigned i;
+
+	if (!drm->debug_bo_front_guard && !drm->debug_bo_back_guard)
+		return;
+
+	VDBG_BO(bo, "front %p back %p\n", bo->guard_front, bo->guard_back);
+
+	if (bo->handle & 1)
+		guard = ~guard;
+
+	if (drm->debug_bo_front_guard)
+		for (i = 0; i < 512; i++)
+			assert(bo->guard_front[i] == guard);
```

Comment:
Is `assert` really the right tool here? Sure, it gets the job done, but it might be a bit confusing, and only works without NDEBUG... I would kind of expect something like VDBG_BO followed by abort here instead...

Reply:
Good point! I'm always forgetting about NDEBUG.